src/btree/row_srch.c

```diff
@@ -116,6 +127,42 @@ __wt_search_insert(
             last_ins = ins;
             key.data = WT_INSERT_KEY(ins);
             key.size = WT_INSERT_KEY_SIZE(ins);
+            /*
+             * We have an optimization to reduce the number of bytes we need to compare during the
+             * search if we know a prefix of the search key matches the keys we have already
+             * compared on the upper stacks. This works because we know the keys become denser down
+             * the stack.
+             *
+             * However, things become tricky if we have another key inserted concurrently to next of
+             * the search key. If we have seen this key inserted upper the stack, we require the
+             * insertion must also be visible to us down the stack. Otherwise, we may wrongly skip
+             * the comparison of a prefix and land on the wrong spot. Here's an example:
+             *
+             * Suppose we have a skip list:
+             *
+             * L1: AA -> BD
```

Comment:
Could you use `BA` as the second key? That way it fully exemplifies the bug, since then we compare A**B** to B**A** and insert it out of order.

Reply:
Good point. That's a mistake. Fixed